tool/tsh/common/putty_config_windows.go

```diff
@@ -0,0 +1,365 @@
+		return trace.Wrap(err)
+	}
+	if err := registry.WriteString(pk, "ProxyHost", sessionStrings.ProxyHost); err != nil {
+		return trace.Wrap(err)
+	}
+	if err := registry.WriteString(pk, "ProxyUsername", sessionStrings.ProxyUsername); err != nil {
+		return trace.Wrap(err)
+	}
+	if err := registry.WriteString(pk, "ProxyTelnetCommand", sessionStrings.ProxyTelnetCommand); err != nil {
+		return trace.Wrap(err)
+	}
+	if err := registry.WriteString(pk, "PublicKeyFile", sessionStrings.PublicKeyFile); err != nil {
+		return trace.Wrap(err)
+	}
+	if err := registry.WriteString(pk, "DetachedCertificate", sessionStrings.DetachedCertificate); err != nil {
+		return trace.Wrap(err)
+	}
+	if err := registry.WriteString(pk, "UserName", sessionStrings.UserName); err != nil {
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+// addHostCAPublicKey adds a host CA to the registry with a set of space-separated hostnames
+func addHostCAPublicKey(keyName string, publicKey string, hostname string) error {
+	registryKeyName := fmt.Sprintf(`%v\%v`, puttyRegistrySSHHostCAsKey, keyName)
+
+	// get the subkey with the host CA key name
+	registryKey, err := registry.GetOrCreateRegistryKey(registryKeyName)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	hostList, _, err := registryKey.GetStringsValue("MatchHosts")
+	if err != nil {
+		// ERROR_FILE_NOT_FOUND is an acceptable error, meaning that the value does not already
+		// exist and it must be created
+		if err != syscall.ERROR_FILE_NOT_FOUND {
+			log.Debugf("Can't get registry value %v: %T", registryKeyName, err)
+			return trace.Wrap(err)
+		}
+	}
+	// initialize an empty hostlist if there isn't one stored under the registry key
+	if len(hostList) == 0 {
+		hostList = []string{}
+	}
+	defer registryKey.Close()
+
+	// add the new hostname to the existing hostList from the registry key (if one exists)
+	hostList = puttyhosts.AddHostToHostList(hostList, hostname)
+
+	// write strings to subkey
+	if err := registry.WriteMultiString(registryKey, "MatchHosts", hostList); err != nil {
+		return trace.Wrap(err)
+	}
+	if err := registry.WriteString(registryKey, "PublicKey", publicKey); err != nil {
+		return trace.Wrap(err)
+	}
+
+	// write dwords for signature acceptance
+	if err := registry.WriteDword(registryKey, "PermitRSASHA1", puttyPermitRSASHA1); err != nil {
+		return trace.Wrap(err)
+	}
+	if err := registry.WriteDword(registryKey, "PermitRSASHA256", puttyPermitRSASHA256); err != nil {
+		return trace.Wrap(err)
+	}
+	if err := registry.WriteDword(registryKey, "PermitRSASHA512", puttyPermitRSASHA512); err != nil {
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+// formatLocalCommandString replaces placeholders in a constant with actual values
+func formatLocalCommandString(tshPath string, cluster string) string {
+	// replace the placeholder "%cluster" with the actual cluster name as passed to the function
+	clusterString := strings.ReplaceAll(puttyProxyTelnetCommand, "%cluster", cluster)
```

Comment:
This is fine, but did you consider using `text/template` instead?

I don't think it really matters, but most Go devs are likely to be more familiar with the built-in templating syntax rather than the custom `%cluster` syntax you've created here.

Reply:
Good idea. I've rewritten this to use `text/template`, refactored the function into the `puttyhosts` package and also written a test for it.

This also caught the fact that I was replacing the `%tsh` and `%cluster` placeholder values myself, while the others (%proxyhost, %host, %port) will be replaced by PuTTY at runtime. This confused me for a good while; not great given that I wrote the code. The new approach will be much clearer going forward.